src/platforms/ios/iosiaphandler.swift

```diff
@@ -0,0 +1,214 @@
+  // the object in C++, marshalling the object into and out of Swift. Given how brittle that is,
+  // we instead having a shadow product cache in this Swift class, which we use if the user wants
+  // to subscribe.
+  private var productList: [Product] = []
+
+  var transactionUpdates: Task<Void, Never>? = nil
+  var errorCallback: (() -> Void)
+  var successCallback: ((NSString, NSString) -> Void)
+
+  @objc init(errorCallback: @escaping (() -> Void), successCallback: @escaping ((NSString, NSString) -> Void)) {
+    self.errorCallback = errorCallback
+    self.successCallback = successCallback
+    super.init()
+    transactionUpdates = newTransactionListenerTask()
+  }
+
+  deinit {
+    transactionUpdates?.cancel()
+  }
+
+  // MARK: - Getting products
+  @objc func getProducts(with productIdentifiers: Set<String>, productRegistrationCallback: @escaping ((NSString, NSString, NSString, NSString, Double, Int) -> Void), registrationCompleteCallback: @escaping (() ->Void), registrationFailureCallback: @escaping (() ->Void)) async {
+      do {
+        productList = try await Product.products(for: Array(productIdentifiers))
+        InAppPurchaseHandler.logger.info(message: "Saved \(productList.count) products")
+
+        for product in productList {
+
+          let productIdentifier = product.id
+          InAppPurchaseHandler.logger.debug (message: "\(productIdentifier): \(product.displayName)")
+          InAppPurchaseHandler.logger.debug (message: product.description)
+
+          guard let subscription = product.subscription else {
+            InAppPurchaseHandler.logger.error(message: "Product does not include subscription: \(product.id)")
+            assertionFailure()
+            continue
+          }
+
+          let trialDays: Int
+          if await subscription.isEligibleForIntroOffer, let offerType = subscription.introductoryOffer?.type, offerType == .introductory, let introOfferPeriod = subscription.introductoryOffer?.period {
+            trialDays = daysInTrial(for: introOfferPeriod.unit, with: introOfferPeriod.value)
+          } else {
+            trialDays = 0
+          }
+
+          let price = product.displayPrice
+          let currencyCode = product.priceFormatStyle.currencyCode
+
+          let numberOfMonthsInSubscription = monthsInSubscription(for: subscription.subscriptionPeriod.unit, with: subscription.subscriptionPeriod.value)
+          let monthlyPrice: Double = (product.price as NSDecimalNumber).doubleValue / Double(numberOfMonthsInSubscription)
+
+          let numberFormatter = NumberFormatter()
+          numberFormatter.numberStyle = .currency
+          numberFormatter.locale = product.priceFormatStyle.locale
+          guard let monthlyPriceString = numberFormatter.string(from: NSNumber(floatLiteral: monthlyPrice)) else {
+            InAppPurchaseHandler.logger.error(message: "Monthly price could not be converted to a string with numberFormatter: \(monthlyPrice)")
+            assertionFailure()
```

Comment:
What happens if we hit this assert here? Generally speaking, an assert should indicate a failure of the programming logic. Could we get here if there was a misconfiguration in the appstore or Guardian or something? While that would be a serious error on our end, it probably doesn't necessitate app termination or sending us coredumps.

Reply:
Hitting that assert: On a debug build we would crash, on a release build we'd log something but the user would have no impact and the app would continue. There is no way this should ever fail, so an `assertionFailure` feels appropriate here to me.

Reply:
Thank you. I am okay with leaving this `assertionFailure()` in then.